SEMANTICS.md

```diff
@@ -0,0 +1,297 @@
+revert any state changes in the calling contract. The error may then be
+intercepted by the calling contract (for `ReplyOn::Always` and
+`ReplyOn::Error`). _In this case, the messages error doesn't abort the whole
+transaction_
+
+#### Handling the Reply
+
+In order to make use of `submessages`, the calling contract must have an extra
+entry point:
+
+```rust
+#[entry_point]
+pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractError> { }
+
+pub struct Reply {
+    pub id: u64,
+    /// ContractResult is just a nicely serializable version of `Result<T, String>`
+    pub result: ContractResult<SubcallResponse>,
+}
+
+pub struct SubcallResponse {
+    pub events: Vec<Event>,
+    pub data: Option<Binary>,
+}
+```
+
+After the `submessage` is finished, the caller will get a chance to handle the
+result. It will get the original `id` of the subcall so it can switch on how to
+process this, and the `Result` of the execution, both success and error. Note
+that it includes all events returned by the submessage, which applies to native
+sdk modules as well (like Bank) as well as the data returned from below. This
+and the original call id provide all context to continue processing it. If you
+need more state, you must save some local context to the store (under the `id`)
+before returning the `submessage` in the original `execute`, and load it in
+`reply`. We explicitly prohibit passing information in contract memory, as that
+is the key vector for reentrancy attacks, which are a large security surface
+area in Ethereum.
+
+The `reply` call may return `Err` itself, in which case it is treated like the
+caller errored, and aborting the transaction. (Unless, of course, the original
+`execute` was itself called from a `submessage`, meaning just the state changes
+since the call that returned `submessage` will be reverted, and this may be
+handled at a higher level). However, on successful processing, `reply` may
+return a normal `Response`, which will be processed as normal - events added to
+the EventManager, and all `messages` and `submessages` dispatched as described
+above.
+
+The one _critical difference_ with `reply`, is that we _do not drop data_. If
+`reply` returns `data: Some(value)` in the `Response` object, we will overwrite
+the `data` field returned by the caller. That is, if `execute` returns
+`data: Some(b"first thought")` and the `reply` (with all the extra information
+it is privvy to) returns `data: Some(b"better idea")`, then this will be
+returned to the caller of `execute` (either the client or another transaction),
+just as if the original `execute` and returned `data: Some(b"better idea")`. If
+`reply` returns `data: None`, it will not modify any previously set data state.
+If there are multiple submessages all setting this, only the last one is used
+(they all overwrite any previous `data` value).
+
+`Submessages` (and their replies) are all executed before any `messages`. They
```

Comment:
When putting those in ticks, let's use the correct vairable/field names `submessages`/`messages`. Here we can also write it as a plain word:

```suggestion
Submessages (and their replies) are all executed before any `messages`. They
```